core/systems/bluetooth/bluetooth_manager.gd

```diff
@@ -0,0 +1,204 @@
+	return [output[0], exit_code]
+
+
+# Run bluetoothctl with the given arguments and an interactive shell.
+func _run_bluetoothctl_scan(status: String) -> void:
+	var cmd := "bluetoothctl scan " + status + "\n"
+	_queue_cmd(cmd)
+	var on_progress := func(output: Array):
+		for line in output:
+			# Send the user's password if prompted
+			if line.contains("[NEW]"):
+				logger.debug("New device! " + line)
+				continue
+			if line.contains("[CHG]"):
+				logger.debug("Change to device! " + line)
+				continue
+			if line.contains("[DEL]"):
+				logger.debug("Device not present! " + line)
+				continue
+
+	# Pass the callback which will watch our command output
+	await _follow_command(cmd, on_progress)
+
+
+# Waits for the given command to produce some output and executes the given 
+# callback with the output.
+func _follow_command(cmd: String, callback: Callable) -> void:
+	# Signal output: [cmd, output, finished]
+	var out: Array = [""]
+	var finished = false
+	while not finished:
+		while out[0] != cmd:
+			out = await command_progressed 
+		var output := out[1] as Array
+		finished = out[2]
+		callback.call(output)
+		# Clear the inner loop condition to fetch progress again
+		out[0] = "" 
+
+
+func _queue_cmd(cmd: String) -> void:
+	cmd_queue.push_back(cmd)
+
+
+func _thread_process(_delta: float) -> void:
+	if not proc:
+		return
+
+	# Process our command queue
+	_process_command_queue()
+
+	# Read the output from the process
+	var output := proc.read()
+
+	# Return if there is no output from the process
+	if output == "":
+		return
+
+	# Split the output into lines
+	var lines := output.split("\n")
+	current_output.append_array(lines)
+
+	# Print the output of steamcmd, except during login for security reasons
+	for line in lines:
+		logger.debug("bluetoothctl: " + line)
+
+	# Signal when command progress has been made 
+	if current_cmd != "":
+		var out := lines.duplicate()
+		#emit_signal.call_deferred("command_progressed", current_cmd, out, false)
+		command_progressed.emit(current_cmd, out, false)
+
+#	# Signal that a steamcmd prompt is available
+#	if lines[-1].begins_with("[bluetooth]#"):
+#		if state == STATE.READY:
+#			emit_signal.call_deferred("client_ready")
+#		state = STATE.PROMPT
+#		prompt_available.emit()
+#
+#		# If a command was executing, emit its output
+#		if current_cmd == "":
+#			return
+#		var out := current_output.duplicate()
+#		#emit_signal.call_deferred("command_progressed", current_cmd, [], true)
+#		command_progressed.emit(current_cmd, [], true)
+#		#emit_signal.call_deferred("command_finished", current_cmd, out)
+#		command_finished.emit(current_cmd, out)
+#		current_cmd = ""
+#		current_output.clear()
+
+
+# Processes commands in the queue by popping the first item in the queue and 
+# setting our state to EXECUTING.
+func _process_command_queue() -> void:
+	if state != STATE.PROMPT or cmd_queue.size() == 0:
+		return
+	var cmd := cmd_queue.pop_front() as String
```

Comment:
Whenever multiple threads (i.e. the main thread + the bluetooth thread) access the same data, you need to protect it with a `Mutex`. This will prevent race conditions if both threads try to modify the data at the same time.

Reply:
Alright. Take a look at your steam client as I pulled this from there.